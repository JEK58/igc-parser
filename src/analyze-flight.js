const turf = require('@turf/turf');

function analyzeFlight(flight, task) {
  let nextTP = 0;

  let start, finish;

  for (let i = 0; i < flight.length - 1; i++) {
    let fix1 = flight[i].coordinate;
    let fix2 = flight[i + 1].coordinate;

    if (nextTP < 2) {
      let point = task.points[0].observationZone.checkEnter(fix1, fix2);
      if (point) {
        nextTP = 1;
        start = { point, i, secOfDay: flight[i].secOfDay };
      }
    }

    if (nextTP === task.points.length - 1) {
      let point = task.points[nextTP].observationZone.checkEnter(fix1, fix2);
      if (point) {
        nextTP += 1;
        finish = { point, i, secOfDay: flight[i].secOfDay };
        break;
      }
    }

    let point = task.points[nextTP].observationZone.checkEnter(fix1, fix2);
    if (point) {
      nextTP += 1;
    }
  }

  return { start, finish };
}

module.exports = analyzeFlight;
