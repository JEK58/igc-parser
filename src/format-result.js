function formatResult(result) {
  return `Start: ${formatTime(result.start.secOfDay)}\nTime: ${formatTime(result.totalTime)}`;
}

function formatTime(secOfDay) {
  let sec = secOfDay % 60;
  let min = ((secOfDay - sec) / 60) % 60;
  let hour = (secOfDay - sec - min * 60) / 3600;
  return `${hour}:${min}:${sec}`;
}

module.exports = formatResult;
