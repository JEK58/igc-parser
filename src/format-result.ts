export function formatResult(result) {
  return `Start: ${formatTime(result.start.secOfDay)}\n` +
    `Time: ${formatTime(result.totalTime)}\n` +
    `Distance: ${(result.distance)} km\n` +
    `Speed: ${(result.speed)} km/h\n`;
}

export function formatTime(secOfDay) {
  let sec = secOfDay % 60;
  let min = ((secOfDay - sec) / 60) % 60;
  let hour = (secOfDay - sec - min * 60) / 3600;
  return `${hour}:${min}:${sec}`;
}
