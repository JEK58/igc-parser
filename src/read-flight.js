const fs = require('fs');

function readFlight(path) {
  return fs.readFileSync(path, 'utf8')
    .split('\n')
    .filter(line => line[0] === 'B')
    .map(line => line.match(/^B(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})([NS])(\d{3})(\d{2})(\d{3})([EW])/))
    .filter(Boolean)
    .map(match => {
      let secOfDay = parseInt(match[1]) * 3600 + parseInt(match[2]) * 60 + parseInt(match[3]);
      let lat = parseInt(match[4]) + parseInt(match[5]) / 60 + parseInt(match[6]) / 60000;
      let lon = parseInt(match[8]) + parseInt(match[9]) / 60 + parseInt(match[10]) / 60000;
      let coordinate = [lon, lat];
      return { secOfDay, coordinate };
    });
}

module.exports = readFlight;
