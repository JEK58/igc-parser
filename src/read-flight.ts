import * as fs from "fs";

const RE_HFDTE = /^HFDTE(\d{2})(\d{2})(\d{2})/;
const RE_B = /^B(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})([NS])(\d{3})(\d{2})(\d{3})([EW])/;

export interface Fix {
  secOfDay: number,
  time: number,
  coordinate: GeoJSON.Position,
}

export function readFlight(path: string): Fix[] {
  let lines = fs.readFileSync(path, 'utf8').split('\n');

  let date: number | undefined;
  let fixes: Fix[] = [];

  lines.forEach(line => {
    if (date === undefined) {
      date = findDate(line);
    }

    // TODO handle UTC-midnight wraparound
    if (date) {
      let fix = convertLine(line, date);
      if (fix) {
        fixes.push(fix);
      }
    }
  });

  return fixes;
}

function findDate(line: string): number | undefined {
  let match = line.match(RE_HFDTE);
  if (match) {
    let year = parseInt(match[3]) + 2000;
    let month = parseInt(match[2]) - 1;
    let day = parseInt(match[1]);
    return Date.UTC(year, month, day);
  }
}

function convertLine(line: string, date: number): Fix | undefined {
  let match = line.match(RE_B);
  if (!match) return;

  let secOfDay = parseInt(match[1]) * 3600 + parseInt(match[2]) * 60 + parseInt(match[3]);
  let time = date + secOfDay * 1000;

  let lat = parseInt(match[4]) + parseInt(match[5]) / 60 + parseInt(match[6]) / 60000;
  let lon = parseInt(match[8]) + parseInt(match[9]) / 60 + parseInt(match[10]) / 60000;
  let coordinate = [lon, lat];

  return { secOfDay, time, coordinate };
}
