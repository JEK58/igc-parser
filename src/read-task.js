const fs = require('fs');
const xml2js = require('xml-js').xml2js;

const oz = require('./oz');

function readTask(path) {
  let file = fs.readFileSync(path, 'utf8');
  let xml = xml2js(file);
  return convertTask(xml.elements.find(it => it.name === 'Task'));
}

function convertTask(xml) {
  return {
    type: xml.attributes.type,
    aatMinTime: parseInt(xml.attributes.aat_min_time),
    points: xml.elements.map(convertPoint),
  };
}

function convertPoint(xml) {
  let waypoint = convertWaypoint(xml.elements.find(it => it.name === 'Waypoint'));
  let observationZone = convertObservationZone(xml.elements.find(it => it.name === 'ObservationZone'), waypoint.location);

  return {
    type: xml.attributes.type,
    waypoint,
    observationZone,
  };
}

function convertWaypoint(xml) {
  return {
    name: xml.attributes.name,
    altitude: parseFloat(xml.attributes.altitude),
    location: convertLocation(xml.elements.find(it => it.name === 'Location')),
  };
}

function convertLocation(xml) {
  return [parseFloat(xml.attributes.longitude), parseFloat(xml.attributes.latitude)];
}

function convertObservationZone(xml, location) {
  let type = xml.attributes.type;

  if (type === 'Line') {
    let length = parseFloat(xml.attributes.length);
    return new oz.Line(location, length);

  } else if (type === 'Cylinder') {
    let radius = parseFloat(xml.attributes.radius);
    return new oz.Cylinder(location, radius);
  }
}

module.exports = readTask;
