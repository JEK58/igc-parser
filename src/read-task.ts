const fs = require('fs');
const xml2js = require('xml-js').xml2js;
const turf = require('@turf/turf');

const oz = require('./oz');
const Turnpoint = require('./turnpoint').Turnpoint;

export function readTask(path) {
  let file = fs.readFileSync(path, 'utf8');
  let xml = xml2js(file);
  let task = convertTask(xml.elements.find(it => it.name === 'Task'));

  task.points.forEach((tp, i) => {
    let bearing;
    if (i === 0) {
      let locNext = task.points[i + 1].location;
      bearing = turf.bearing(tp.location, locNext);

    } else if (i === task.points.length - 1) {
      let locPrev = task.points[i - 1].location;
      bearing = turf.bearing(locPrev, tp.location);

    } else {
      let locPrev = task.points[i - 1].location;
      let locNext = task.points[i + 1].location;

      let bearingFromPrev = turf.bearing(locPrev, tp.location);
      let bearingToNext = turf.bearing(tp.location, locNext);

      bearing = turf.bearingToAngle((bearingToNext + bearingFromPrev) / 2);
    }

    tp.observationZone.bearing = bearing;

    if (tp.observationZone.update) {
      tp.observationZone.update();
    }
  });

  return task;
}

function convertTask(xml) {
  return {
    type: xml.attributes.type,
    aatMinTime: parseInt(xml.attributes.aat_min_time),
    points: xml.elements.map(convertPoint),
  };
}

function convertPoint(xml) {
  let waypoint = convertWaypoint(xml.elements.find(it => it.name === 'Waypoint'));
  let observationZone = convertObservationZone(xml.elements.find(it => it.name === 'ObservationZone'), waypoint.location);

  return new Turnpoint(waypoint.name, waypoint.altitude, waypoint.location, observationZone);
}

function convertWaypoint(xml) {
  return {
    name: xml.attributes.name,
    altitude: parseFloat(xml.attributes.altitude),
    location: convertLocation(xml.elements.find(it => it.name === 'Location')),
  };
}

function convertLocation(xml) {
  return [parseFloat(xml.attributes.longitude), parseFloat(xml.attributes.latitude)];
}

function convertObservationZone(xml, location) {
  let type = xml.attributes.type;

  if (type === 'Line') {
    let length = parseFloat(xml.attributes.length);
    return new oz.Line(location, length);

  } else if (type === 'Cylinder') {
    let radius = parseFloat(xml.attributes.radius);
    return new oz.Cylinder(location, radius);
  }
}
