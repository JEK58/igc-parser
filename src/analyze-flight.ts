import {Fix} from "./read-flight";
import Task from "./task/task";
import RacingTaskSolver from "./task/solver/racing-task-solver";
import AreaTaskSolver from "./task/solver/area-task-solver";

export function analyzeFlight(flight: Fix[], task: Task) {
  if (task.options.isAAT) {
    let analyzer = new AreaTaskSolver(task);
    analyzer.consume(flight);
    return analyzer.result;
  } else {
    let analyzer = new RacingTaskSolver(task);
    analyzer.consume(flight);
    return analyzer.result;
  }
}
