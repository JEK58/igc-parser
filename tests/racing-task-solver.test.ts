import {readTask} from "../src/read-task";
import RacingTaskSolver from "../src/racing-task-solver";
import Task from "../src/task";
import {Fix, readFlight} from "../src/read-flight";

describe('RacingTaskSolver', () => {
  describe('with task "2017-07-17-lev.tsk"', () => {
    let task: Task;
    let solver: RacingTaskSolver;
    let flight: Fix[];

    beforeEach(() => {
      task = readTask(`${__dirname}/../fixtures/2017-07-17-lev.tsk`);
      solver = new RacingTaskSolver(task);
      flight = readFlight(`${__dirname}/../fixtures/2017-07-17-lev/IGP_77hg7sd1.IGC`);
    });

    it('emits multiple start events', () => {
      let fixes: Fix[] = [];
      solver.on('start', fix => fixes.push(fix));
      flight.forEach(fix => solver.update(fix));
      expect(fixes).toMatchSnapshot();
    });

    it('emits multiple turn events', () => {
      let fixes = [];
      solver.on('turn', (fix, i) => fixes.push([fix, i]));
      flight.forEach(fix => solver.update(fix));
      expect(fixes).toMatchSnapshot();
    });

    it('emits single finish event', () => {
      let fixes: Fix[] = [];
      solver.on('finish', fix => fixes.push(fix));
      flight.forEach(fix => solver.update(fix));
      expect(fixes).toMatchSnapshot();
    });

    it('returns a result', () => {
      flight.forEach(fix => solver.update(fix));
      expect(solver.result).toMatchSnapshot();
    });

    it('can handle outlandings', () => {
      flight = readFlight(`${__dirname}/../fixtures/2017-07-17-lev/ZG_77hv6ci1.igc`);
      flight.forEach(fix => solver.update(fix));
      expect(solver.result).toMatchSnapshot();
    })
  });
});
