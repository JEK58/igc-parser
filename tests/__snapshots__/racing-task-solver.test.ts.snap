// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" can handle outlandings 1`] = `
Object {
  "completed": false,
  "distance": 94199.64354856145,
  "finish": undefined,
  "time": undefined,
  "turns": Array [
    Object {
      "point": Array [
        6.28125,
        50.989016666666664,
      ],
      "time": 1500306180000,
    },
    Object {
      "point": Array [
        6.1937500000000005,
        51.1723,
      ],
      "time": 1500308465000,
    },
  ],
  "validStarts": Array [
    Object {
      "point": Array [
        6.926316666666667,
        51.10086666666667,
      ],
      "time": 1500299777000,
    },
    Object {
      "point": Array [
        6.926533333333333,
        51.10013333333333,
      ],
      "time": 1500299800000,
    },
    Object {
      "point": Array [
        6.926566666666667,
        51.09955,
      ],
      "time": 1500299823000,
    },
    Object {
      "point": Array [
        6.9270000000000005,
        51.09813333333334,
      ],
      "time": 1500299892000,
    },
    Object {
      "point": Array [
        6.9177333333333335,
        51.1278,
      ],
      "time": 1500300293000,
    },
    Object {
      "point": Array [
        6.9177,
        51.12821666666667,
      ],
      "time": 1500300324000,
    },
    Object {
      "point": Array [
        6.917783333333333,
        51.127583333333334,
      ],
      "time": 1500300351000,
    },
    Object {
      "point": Array [
        6.917766666666667,
        51.12786666666667,
      ],
      "time": 1500300386000,
    },
    Object {
      "point": Array [
        6.91795,
        51.127633333333335,
      ],
      "time": 1500300415000,
    },
    Object {
      "point": Array [
        6.92185,
        51.11451666666667,
      ],
      "time": 1500301090000,
    },
    Object {
      "point": Array [
        6.922066666666667,
        51.11431666666667,
      ],
      "time": 1500301112000,
    },
    Object {
      "point": Array [
        6.922016666666667,
        51.11431666666667,
      ],
      "time": 1500301132000,
    },
    Object {
      "point": Array [
        6.928533333333333,
        51.093700000000005,
      ],
      "time": 1500302758000,
    },
    Object {
      "point": Array [
        6.930083333333334,
        51.087700000000005,
      ],
      "time": 1500302925000,
    },
    Object {
      "point": Array [
        6.93065,
        51.08641666666667,
      ],
      "time": 1500303226000,
    },
    Object {
      "point": Array [
        6.930366666666667,
        51.08711666666667,
      ],
      "time": 1500303283000,
    },
  ],
}
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" emits multiple start events 1`] = `
Array [
  Object {
    "coordinate": Array [
      6.914216666666667,
      51.13826666666667,
    ],
    "secOfDay": 51431,
    "time": 1500301031000,
  },
  Object {
    "coordinate": Array [
      6.92305,
      51.10805,
    ],
    "secOfDay": 52011,
    "time": 1500301611000,
  },
  Object {
    "coordinate": Array [
      6.92745,
      51.096900000000005,
    ],
    "secOfDay": 53015,
    "time": 1500302615000,
  },
  Object {
    "coordinate": Array [
      6.930316666666667,
      51.08435,
    ],
    "secOfDay": 53459,
    "time": 1500303059000,
  },
]
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" emits multiple turn events 1`] = `
Array [
  Array [
    Object {
      "coordinate": Array [
        6.2829,
        50.996633333333335,
      ],
      "secOfDay": 55467,
      "time": 1500305067000,
    },
    1,
  ],
  Array [
    Object {
      "coordinate": Array [
        6.194816666666667,
        51.173249999999996,
      ],
      "secOfDay": 57115,
      "time": 1500306715000,
    },
    2,
  ],
  Array [
    Object {
      "coordinate": Array [
        6.6080499999999995,
        50.96866666666667,
      ],
      "secOfDay": 58807,
      "time": 1500308407000,
    },
    3,
  ],
]
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" emits single finish event 1`] = `
Array [
  Object {
    "coordinate": Array [
      7.008533333333333,
      51.09851666666667,
    ],
    "secOfDay": 59951,
    "time": 1500309551000,
  },
]
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" returns a result 1`] = `
Object {
  "completed": true,
  "distance": 136420.85415876907,
  "finish": Object {
    "point": Array [
      7.008533333333333,
      51.09851666666667,
    ],
    "time": 1500309551000,
  },
  "time": 6492,
  "turns": Array [
    Object {
      "point": Array [
        6.2829,
        50.996633333333335,
      ],
      "time": 1500305067000,
    },
    Object {
      "point": Array [
        6.194816666666667,
        51.173249999999996,
      ],
      "time": 1500306715000,
    },
    Object {
      "point": Array [
        6.6080499999999995,
        50.96866666666667,
      ],
      "time": 1500308407000,
    },
  ],
  "validStarts": Array [
    Object {
      "point": Array [
        6.914216666666667,
        51.13826666666667,
      ],
      "time": 1500301031000,
    },
    Object {
      "point": Array [
        6.92305,
        51.10805,
      ],
      "time": 1500301611000,
    },
    Object {
      "point": Array [
        6.92745,
        51.096900000000005,
      ],
      "time": 1500302615000,
    },
    Object {
      "point": Array [
        6.930316666666667,
        51.08435,
      ],
      "time": 1500303059000,
    },
  ],
}
`;
